### PR TITLE
Try to avoid multiple agents stealing one dead agent's work

### PR DESCRIPTION
**Pull request by sfc-gh-anoyes**
Previously we were reading at snapshot isolation when deciding whether or not we can steal work from an agent who hasn't heartbeated recently. If several agents are idle when an agent stops heartbeating, they might all succeed in stealing its work, which can result in overshooting max_runs. Now we add a read conflict which should allow at most one agent to steal work from a presumed dead agent.

### CODE REVIEW

File: joshua/joshua_model.py [hunk 1]
```diff
@@ -621,7 +628,12 @@ def should_run_ensemble(tr : fdb.Transaction, ensemble_id : str) -> bool:
             # No other agents are running a test for this ensemble (is this possible?)
             return True
         if max_heartbeat_age > 10:
-            # "Kill" this seed for the presumed dead agent
+            print("Agent {} presumed dead. Attempting to steal its work.".format(
+                _get_hostname(ensemble_id, max_seed, tr)))
+
+            # If we read at snapshot isolation then an arbitrary number of agents could steal this run/seed.
+            # We only want one agent to succeed in taking over for the dead agent's run/seed.
+            tr.add_read_conflict_key(dir_ensemble_incomplete[ensemble_id]["heartbeat"][max_seed])
```
Comment (sfc-gh-kmakino) [inline joshua/joshua_model.py:636]: Is it OK to rely on the snapshotted property above in line 615?
Don't we want to make sure the deletion here will always be visible?
```
props = _get_ensemble_properties(tr, ensemble_id, snapshot=True)
```

Comment (sfc-gh-kmakino) [reply joshua/joshua_model.py:636]: > Is it OK to rely on the snapshotted property above in line 615?
> Don't we want to make sure the deletion here will always be visible?
> 
> ```
> props = _get_ensemble_properties(tr, ensemble_id, snapshot=True)
> ```

Never mind, I was a bit confused but @sfc-gh-anoyes explained to me offline.

**Review by sfc-gh-kmakino — APPROVED** [review]
LGTM!

File: joshua/joshua_model.py (hunk 1 above)
Comment (sfc-gh-anoyes) [reply joshua/joshua_model.py:636]: Discussed offline - snapshot isolation doesn't have to do with when mutations are visible

**Pull request merged**